clarity/src/vm/analysis/type_checker/tests/contracts.rs

```diff
@@ -630,3 +664,1864 @@ fn test_expects() {
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-1 (contract <trait-1>))
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+
+        let trait_to_compatible_trait = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-1>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-2>))
+            (ok true))";
+
+        let bad_principal_to_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-1 (contract principal))
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+
+        let bad_other_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-2 (contract <trait-1>))
+            (internal-get-2 contract))
+        (define-public (internal-get-2 (contract <trait-2>))
+            (contract-call? contract get-2 u1))";
+
+        let embedded_trait = "(define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-12>))
+            (internal-get-1 (some contract)))
+        (define-public (internal-get-1 (opt-contract (optional <trait-12>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
+
+        let embedded_trait_compatible = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-12>))
+            (internal-get-1 (some contract)))
+        (define-public (internal-get-1 (opt-contract (optional <trait-1>)))
+            (match opt-contract
+                contract (contract-call? contract get-1 u1)
+                (err u1)
+            )
+        )";
+
+        let bad_embedded_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-opt-get-1 (contract <trait-1>))
+            ;; Passing (optional <trait-1>) as an (optional <trait-12>) should be an error
```

Comment:
But the other way around should work, right?  Passing an `(optional <triat-12>)` to a function that takes `(optional <trait-1>)` should work?

Reply:
Yes, that is verified in the test above this, in what has now been split out into `test_embedded_trait_compatible`.